Std/Data/Array/Basic.lean

```diff
@@ -117,6 +117,102 @@ protected def maxI [ord : Ord α] [Inhabited α]
   xs.minI (ord := ord.opposite) start stop
 
+
+/--
+Returns the first and all the rest of the elements of the array.
+-/
+protected def splitFirst (xs : Array α) (h : 0 < xs.size) : α × Array α :=
+  let head := xs.get ⟨0, h⟩
+  let tail := Subarray.mk xs 1 xs.size (Nat.succ_le_of_lt h) (Nat.le_refl _) |>.toArray
+  (head, tail)
+
+/--
+Returns the first and all the rest of the elements of the array, or `none` if it is empty.
+-/
+protected def splitFirst? (xs : Array α) : Option (α × Array α) :=
+  if h : 0 < xs.size then
+    let head := xs.get ⟨0, h⟩
+    let tail := Subarray.mk xs 1 xs.size (Nat.succ_le_of_lt h) (Nat.le_refl _) |>.toArray
+    some (head, tail)
+  else
+    none
+
+/--
+Returns the last and all the rest of the elements of the array, or `none` if it is empty.
+-/
+protected def splitLast? (xs : Array α) : Option (Array α × α) :=
+  if h1 : 0 < xs.size then
+    have h2 := Nat.sub_lt h1 (by decide)
+    let last := xs.get ⟨xs.size - 1, h2⟩
+    let heads := Subarray.mk xs 0 (xs.size - 1) (Nat.zero_le _) (Nat.le_of_lt h2) |>.toArray
+    some (heads, last)
+  else
+    none
+
+/--
+Returns the last and all the rest of the elements of the array.
+-/
+protected def splitLast (xs : Array α) (h : 0 < xs.size) : α × Array α :=
+  have h2 := Nat.sub_lt h (by decide)
+  let heads := xs.get ⟨xs.size - 1, h2⟩
+  let last := Subarray.mk xs 0 (xs.size - 1) (Nat.zero_le _) (Nat.le_of_lt h2) |>.toArray
+  (heads, last)
+
+/--
+Returns the last element of the array, or `none` if it is empty.
+-/
+protected def last? (xs : Array α) : Option α :=
+  if h : 0 < xs.size then
+    some (xs.get ⟨xs.size - 1, Nat.sub_lt h (by decide)⟩)
+  else
+    none
+
+/--
+Returns the last element of the array.
+-/
+protected def last (xs : Array α) (h : 0 < xs.size) : α :=
+  (xs.get ⟨xs.size - 1, Nat.sub_lt h (by decide)⟩)
+
+/--
+Returns the first element of the array, or `none` if it is empty.
+-/
+protected def first? (xs : Array α) : Option α :=
+  if h : 0 < xs.size then
+    some (xs.get ⟨0, h⟩)
+  else
+    none
+
+/--
+Returns the first element of the array.
+-/
+protected def first (xs : Array α) (h : 0 < xs.size) : α :=
+  xs.get ⟨0, h⟩
+
+/--
+Split an array apart into two arrays at a given index. The first half
+will contain all values from `[0, mid)`, the second will contain `[mid, xs.size)`.
+-/
+protected def splitAt (xs : Array α) (mid : Fin xs.size) : Array α × Array α :=
+  have h := (Nat.le_of_lt mid.isLt)
+  let lhs := Subarray.mk xs 0 mid.val (Nat.zero_le _) h |>.toArray
+  let rhs := Subarray.mk xs mid.val xs.size h (Nat.le_refl _) |>.toArray
+  (lhs, rhs)
+
+/--
+Split an array apart into two arrays at a given index. The first half
+will contain all values from `[0, mid)`, the second will contain `[mid, xs.size)`.
+Since `mid` is not provably an index of `xs` this variant will return `none`
+in case it isn't.
+-/
+protected def splitAt? (xs : Array α) (mid : Nat) : Option (Array α × Array α) :=
```

Comment:
I'm not sure an option-returning version of this function is useful. More useful would be to just return `(xs, #[])` when `mid` is out of range.